streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordCollectorImpl.java

```diff
@@ -146,17 +192,22 @@ public void onCompletion(final RecordMetadata metadata,
                 "its internal buffer fills up. " +
                 "You can increase producer parameter `max.block.ms` to increase this timeout.", topic);
             throw new StreamsException(String.format("%sFailed to send record to topic %s due to timeout.", logPrefix, topic));
-        } catch (final Exception fatalException) {
-            throw new StreamsException(
-                String.format(EXCEPTION_MESSAGE,
-                              logPrefix,
-                              "an error caught",
-                              key,
-                              value,
-                              timestamp,
-                              topic,
-                              fatalException.getMessage()),
-                fatalException);
+        } catch (final Exception uncaughtException) {
```

Comment:
`KafkaProducer.send()` can throw the following:
```
     * @throws IllegalStateException if a transactional.id has been configured and no transaction has been started
     * @throws InterruptException If the thread is interrupted while blocked
     * @throws SerializationException If the key or value are not valid objects given the configured serializers
     * @throws TimeoutException If the time taken for fetching metadata or allocating memory for the record has surpassed <code>max.block.ms</code>.
     * @throws KafkaException If a Kafka related error occurs that does not belong to the public API exceptions.
```
I think we need to stop the world for all of them and thus don't need to call the handler.

Maybe we should specify in the KIP, that the handler only applies to errors reported in the `Callback` but not to exceptions throws by `send()` directly ? \cc @guozhangwang @bbejeck @dguy WDYT?

Reply:
I'd be fine with that change — I thought there was a more diverse set of things that might fire back immediately when I did my first read-through.